### PR TITLE
feat: adjust date time property conventions rule (#1536)

### LINKED ISSUE #1501: Relax naming convention for date/time properties
Please check if the PR https://github.com/zalando/restful-api-guidelines/pull/811 introduces changes which are relevant to the Zally project.

### PR DESCRIPTION
**Pull request by tkrop**
This pull request adjusts the date time property suffix rule according to the rule change conducted in https://github.com/zalando/restful-api-guidelines/pull/811.

fixes #1501.

### CODE REVIEW

File: server/zally-ruleset-zalando/src/main/resources/reference.conf [hunk 1]
```diff
@@ -216,7 +216,11 @@ ProprietaryHeadersRule {
 
 DateTimePropertiesSuffixRule {
   patterns: [
-    ".*_at"
+    ".*_at",
+    ".*_time",
+    ".*_timestamp",
+    ".*_date",
+    ".*_day",
```
Comment (SmilyOrg) [inline server/zally-ruleset-zalando/src/main/resources/reference.conf:223]: Hmm, the rule says _contains_, but this needs it to be a suffix, should these be something like `".*_time(_.*)?` instead?

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Valid point, but I think it requires `.*_time$` instead, since it is using a regular expression matcher and there is no default `$` added to the pattern. I will create a regression test that proofs whether the current implementations is incorrect and how to fix it.

Thanks for making me think.

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Confirmed. The current implementation is for some reason only matching suffixes and fails for property patterns like `<my-var>_[time,time-date,date,day]_suffix`.

However, i'm also getting the impression that a `contains` does not make sense. We neither have we listed an example nor can I find one.

Comment (ePaul) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: I could think of things like "estimated_time_of_arrival", which has the `time` in the middle.

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Makes sense. I will try to adjust the regular expressions and tests to support the contains approach. Do we actually also `time`..., etc. to be positioned as prefix?

Comment (SmilyOrg) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Just `time_of_arrival` might also make sense, in which case it would be a prefix. Maybe the check could be that it uses one of these words with an underscore on at least one side. Just `date`, `time` alone is probably not very useful.

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: I fear, the last one is hard to express in regular expression, but I think that `(.*_)*<key>(_.*)*` would be an appropriate prefix/infix/suffix regular expression for our case.

Comment (SmilyOrg) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: What about https://regex-vis.com/?r=%28.%2B_date%28_.%2B%29%3F%7C%28.%2B_%29%3Fdate_.%2B%29&t=%5B%22date%22%2C%22date_of_arrival%22%2C%22creation_date%22%2C%22_date%22%2C%22date_%22%2C%22estimated_date_of_arrival%22%5D?

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: I think that `(.+_)*<key>(_.+)*` is a simplified 100% equivalent version.

Comment (SmilyOrg) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Almost, but not 100%. It's a bit too permissive for edge cases, but if we don't care about them, then it's also good.
image

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: Okay, convinced. I'm using your expression.

**Comment by ePaul** [comment]
:+1:

File: server/zally-ruleset-zalando/src/main/resources/reference.conf (hunk 1 above)
Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: @SmilyOrg following your suggestion I see the disadvantage of the readability of the supported patterns in the error message:

```
[ .+_at, (.+_time(stamp)?(_.+)?|(.+_)?time(stamp)?_.+), (.+_date(_.+)?|, .+_)?date_.+), (.+_day(_.+)?|(.+_)?day_.+) ]
```

than

```
[ .+_at, (.+_)*time(_.+)*, (.+_)*time(stamp)?(_.+)*, (.+_)*date(_.+)*, (.+_)*date(_.+)* ]
```

Comment (tkrop) [reply server/zally-ruleset-zalando/src/main/resources/reference.conf]: After extending the test to cover all edge-cases, I can say that both alternatives are equivalent. So I choose the simpler looking version.

File: server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/DateTimePropertiesSuffixRule.kt [hunk 2]
```diff
@@ -13,7 +13,7 @@ import org.zalando.zally.rule.api.Violation
     ruleSet = ZalandoRuleSet::class,
     id = "235",
     severity = Severity.SHOULD,
-    title = "Name date/time properties using the \"_at\" suffix"
+    title = "Name date/time properties using the common suffix"
```
Comment (SmilyOrg) [inline server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/DateTimePropertiesSuffixRule.kt]:
```suggestion
    title = "Name date/time properties using the common conventions"
```

**Review by SmilyOrg — APPROVED** [review]

File: server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/DateTimePropertiesSuffixRuleTest.kt [hunk 3]
```diff
@@ -54,25 +56,58 @@ class DateTimePropertiesSuffixRuleTest {
...
         val violations = rule.validate(DefaultContextFactory().getOpenApiContext(content))
-        assertThat(violations).isEmpty()
+        assertThat(violations.map { it.description }).containsExactly(
+            rule.generateMessage("was_created_at_suffix", "string", "date-time")
+        )
     }
 
     @Test
-    fun `should ignore fields with non-date and time types`() {
+    fun `should fail on prefix _at and succeed on prefix patterns`() {
```
Comment (ePaul) [inline server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/DateTimePropertiesSuffixRuleTest.kt]:
```suggestion
    fun `should fail on prefix at_ and succeed on prefix patterns`() {
```

**Comment by tkrop** [comment]
:+1:

**Comment by ePaul** [comment]
:+1:
